### PR TITLE
[FIX] web_editor: table format breaks on enter key press

### PR DESCRIPTION
**Pull request by dhba-odoo**
**Current behavior before PR:**

- Pressing Enter after selecting all columns in a row would break the table
  format.

**Desired behavior after PR is merged:**
- Now, when the Enter key is pressed, all the contents of the anchor `<td>`
  will be removed first, and then the Enter process within that `<td>` will
  be executed.

task-3557503

### CODE REVIEW

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 1]
```diff
@@ -790,14 +790,15 @@ export function getDeepRange(editable, { range, sel, splitText, select, correctT
...
     const endLeaf = firstLeaf(end);
     const beforeEnd = endLeaf.previousSibling;
+    const previous = previousLeaf(endLeaf, editable, true);
     if (
         correctTripleClick &&
         !endOffset &&
         (start !== end || startOffset !== endOffset) &&
-        (!beforeEnd || (beforeEnd.nodeType === Node.TEXT_NODE && !isVisibleStr(beforeEnd)))
+        (!beforeEnd || (beforeEnd.nodeType === Node.TEXT_NODE && !isVisibleStr(beforeEnd))) &&
+        previous && !closestElement(previous, 'table')
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:799]: Why can't the leaf before the one at the end of the range be in a table?

---

Also, you might as well put the contents of the nested `if` in the top `if` then while you're at it.

```suggestion
        previous && !closestElement(previous, 'table') && closestElement(previous).isContentEditable)
    ) {
```

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js [hunk 2]
```diff
@@ -3830,6 +3830,88 @@ X[]
                     contentAfter: '<p><br></p><p>[]<br></p>',
                 });
             });
+            it('preserve selected table row and insert paragraph in cell when whole table is not selected and enter key is pressed', async () => {
+                // Forward selection
+                await testEditor(BasicEditor, {
+                    contentBefore: unformat(
+                        `<table class="o_selected_table">
+                            <tbody>
+                                <tr><td class="o_selected_td"><p>[<br></p></td><td class="o_selected_td"><p><br></p></td><td class="o_selected_td"><p><br>]</p></td></tr>
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js:3839]: Do you have to put the classes? Doesn't the editor do it on its own when its sets the selection? Ideally you'd have `contentBefore` without the classes, and `contentBeforeEdit` would be what you wrote here.

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js [hunk 3]
```diff
@@ -3830,6 +3830,88 @@ X[]
...
+                                <tr><td><p><br></p></td><td><p><br></p></td><td><p><br></p></td></tr>
+                            </tbody>
+                        </table>`,
+                    ),
+                    stepFunction: insertParagraphBreak,
+                    contentAfter: unformat(
+                        `<table>
+                            <tbody>
+                                <tr><td><p><br></p><p>[]<br></p></td><td><p><br></p></td><td><p><br></p></td></tr>
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js:3848]: ~~Very strange, when I tested manually I got the extra `p` in the _last_ `td` (the one in which the selection ended, like on GDocs if I'm not mistaken). How come this is different?~~

_Woaw I went and tested on Chrome (I'm on Firefox normally) and the behavior matches what you wrote here. So I got my answer_ :shrug:

UPDATE: I went and tested again and noticed that it works only once per row: ENTER will add a `p` in the last `td`, then select again, press ENTER again, and no extra `p` gets added.

**Review by Zynton — CHANGES_REQUESTED** [review]
Nice job.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 4]
```diff
@@ -2258,7 +2258,13 @@ export class OdooEditor extends EventTarget {
                 }
                 return;
             }
-            this.deleteRange(sel);
+            const table = closestElement(sel.anchorNode, 'table');
+            // Checks if all cells within the table are selected.
+            const isTableFullySelected = table && [...table.querySelectorAll('td')]
+                .every(td => td.classList.contains('o_selected_td'));
+            if (BACKSPACE_ONLY_COMMANDS.includes(method) || isTableFullySelected || !closestElement(sel.anchorNode, 'td.o_selected_td')) {
```
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: I need this condition to check that only the text inside a column is selected and not the column. In such a scenario, when the enter key is pressed, the selected text should be removed, and a new paragraph tag will be inserted. 
if removed this condition, only a new paragraph tag will be inserted.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 5]
```diff
@@ -2258,7 +2258,13 @@ export class OdooEditor extends EventTarget {
                 }
                 return;
             }
-            this.deleteRange(sel);
+            const table = closestElement(sel.anchorNode, 'table');
+            // Checks if all cells within the table are selected.
+            const isTableFullySelected = table && [...table.querySelectorAll('td')]
+                .every(td => td.classList.contains('o_selected_td'));
```
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: done!

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js (hunk 2 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js]: do you mean i need to use `contentBeforeEdit`  here?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 6]
```diff
@@ -2252,6 +2253,7 @@ export class OdooEditor extends EventTarget {
         this._resetLinkZws();
         if (!sel.isCollapsed && BACKSPACE_FIRST_COMMANDS.includes(method)) {
             let range = getDeepRange(this.editable, {sel, splitText: true, select: true, correctTripleClick: true});
+            const isBackSpaceCommand = BACKSPACE_ONLY_COMMANDS.includes(method);
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2256]: No need for this, I found it more explicit before.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 7]
```diff
@@ -2260,14 +2262,16 @@ export class OdooEditor extends EventTarget {
...
-                if (BACKSPACE_ONLY_COMMANDS.includes(method)) {
+                if (isBackSpaceCommand) {
                     this._applyRawCommand(method);
                 }
                 return;
             }
-            this.deleteRange(sel);
-            if (BACKSPACE_ONLY_COMMANDS.includes(method)) {
-                return true;
+            if (isBackSpaceCommand || hasTableFullySelected(range) || !closestElement(sel.anchorNode, 'td.o_selected_td')) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2270]:
```suggestion
            if (
                BACKSPACE_ONLY_COMMANDS.includes(method) ||
                isTableFullySelected(range) ||
                closestElement(sel.anchorNode, 'td.o_selected_td')
            ) {
```
Please also add a comment, in particular to explain what you explained to me regarding the third condition.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 8]
```diff
@@ -2260,14 +2262,16 @@ export class OdooEditor extends EventTarget {
...
+                this.deleteRange(sel);
+                if (isBackSpaceCommand) {
+                    return true;
+                }
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2274]: This should remain out of the `if` for clarity.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 9]
```diff
@@ -456,6 +456,19 @@ export function getAdjacents(node, predicate = n => !!n) {
 export function hasTableSelection(editable) {
     return !!editable.querySelector('.o_selected_table');
 }
+/**
+ * Checks if the entire table associated with the given range is fully selected.
+ *
+ * @param {Range} range
+ * @returns {boolean} - True if the entire table is fully selected, false otherwise.
+ */
+export function hasTableFullySelected(range) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:465]: Please rename to `isTableFullySelected`. Also, would this work if the range is not deep? If not, then your function should start by calling `getDeepRange` to replace the range passed as argument.

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js (hunk 2 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js]: Yes. `contentBefore` without the `o_selected_*` classes, `contentBeforeEdit` with the `o_selected_*` classes.

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js [hunk 10]
```diff
@@ -3839,6 +3839,88 @@ X[]
                     contentAfter: '<p><br></p><p>[]<br></p>',
                 });
             });
+            it('preserve selected table row and insert paragraph in cell when whole table is not selected and enter key is pressed', async () => {
+                // Forward selection
+                await testEditor(BasicEditor, {
+                    contentBefore: unformat(
+                        `<table>
+                            <tbody>
+                                <tr><td><p>[<br></p></td><td><p><br></p></td><td><p><br>]</p></td></tr>
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js:3848]: If you're going to use `unformat` anyway, you might as well format your HTML so it doesn't have a 103-character line that is hard to parse at a glance. One line per `td` would be good.

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js (hunk 3 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js:3857]: Have you solved that part? If so please add tests.

**Review by Zynton — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js (hunk 3 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js:3857]: How do you produce it? It's working properly for me. As you can see in the [video](https://www.awesomescreenshot.com/video/24664514?key=2b2e625e2b1a267f63a1b9898054fa49)

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js:3857]: My bad, I just tried and when I thought it wasn't working it was because I was selecting in the opposite direction so the paragraph was added in a different cell and that didn't make the row grow. However, I would have expected the selection to be at the end of the cell after adding the paragraph, and [here](https://www.awesomescreenshot.com/video/24702990?key=84d46cbbc34f7f29450bc1e4c578d65e) it was in the middle of it.

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js:3857]: This issue occurs because the selection shifts to the 'tr' when multiple rows are selected.
When i try to set the selection at that time, the functionality of selecting table stopped working properly.
so, to address this problem, we need to change how table is being selected. Would you like me to proceed with it in this task?

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js:3857]: It depends on what you mean with "change how table is being selected"... The feature should behave as expected so, yes, but that shouldn't mean rewriting the whole feature. Try to make this work in a separate commit so we can see at what cost the fix comes.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 7 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: Here's what you explained to me regarding the third condition:
image
I'd like, at the minimum, that explanation (nicely phrased for the reader of course) to be included in the comment.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 11]
```diff
@@ -2281,7 +2282,19 @@ export class OdooEditor extends EventTarget {
                 }
                 return;
             }
-            this.deleteRange(sel);
+            /**
+             * This condition evaluates three potential scenarios:
+             * 1. The action being performed is a "backspace only" command.
+             * 2. The entire table is selected.
+             * 3. The current selection is not a table selection.
+             */
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2290]: When I asked you to write a comment, I meant one that would explain _why_ that condition is there, not _what_ the condition is (the code is self explanatory in that regard).

In particular, I had asked you to include an explanation regarding the third condition, as this is the one that was neast obvious to me and you had explained it in the review process:
image

But please include an explanation for all three.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 12]
```diff
@@ -2414,22 +2414,19 @@ export class OdooEditor extends EventTarget {
...
-            // We need the triple click correction only for a bug in firefox
-            // where it gives a selection of a full cell as tr 0 tr 1. The
-            // correction makes it so it gives us the cell and not its neighbor.
-            // In all other cases we don't want to make that correction so as to
-            // avoid flicker when hovering borders.
-            range = getDeepRange(this.editable, { correctTripleClick: anchorNode && anchorNode.nodeName === 'TR' });
+        if (this.isFirefox) {
+            if (selection.rangeCount > 1) {
+                this.historyResetLatestComputedSelection(true);
+            } else if(ev && closestElement(ev.target, 'table') === closestElement(selection.anchorNode, 'table')) {
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2420]: What is the idea with this condition?

```suggestion
            } else if (ev && closestElement(ev.target, 'table') === closestElement(selection.anchorNode, 'table')) {
```

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 13]
```diff
@@ -2414,22 +2414,19 @@ export class OdooEditor extends EventTarget {
...
-        } else {
-            // We need the triple click correction only for a bug in firefox
-            // where it gives a selection of a full cell as tr 0 tr 1. The
-            // correction makes it so it gives us the cell and not its neighbor.
-            // In all other cases we don't want to make that correction so as to
-            // avoid flicker when hovering borders.
-            range = getDeepRange(this.editable, { correctTripleClick: anchorNode && anchorNode.nodeName === 'TR' });
+        if (this.isFirefox) {
+            if (selection.rangeCount > 1) {
+                this.historyResetLatestComputedSelection(true);
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2419]: How does this work?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 14]
```diff
@@ -2414,22 +2414,19 @@ export class OdooEditor extends EventTarget {
...
+            } else if(ev && closestElement(ev.target, 'table') === closestElement(selection.anchorNode, 'table')) {
+                setSelection(selection.anchorNode, selection.anchorOffset, ev.target, 0);
```
Comment (Zynton) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2421]: We used to set a range, not a selection because we call `_handleSelectionInTable` in `_onSelectionChange` so here you're causing a risk of infinite loop! Same is true with your call to `historyResetLatestComputedSelection` above.

**Review by Zynton — CHANGES_REQUESTED** [review]

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 14 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2421]: I've checked that these changes do not cause issues like infinite loops, and I haven't encountered any such problems. As I mentioned before, in Firefox, the selection shifts to the `tr` when selecting mutiple `tds`, so if I don't set the selection back properly, the cursor will always come to the second `p`, as `getdeepRange` in `_applyRawCommand` will set the selection at the first `p` of the last `td` before `oEnter` gets called. 

If you believe it's not advisable to set the selection here, then another option would be to set the selection based on the selected `tds` at the time of `mouseup`. However, at that time, it's difficult to decide which `td` should be the anchor node and which should be the focus node.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 13 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2419]: As I have mentioned before, when we select multiple `tds` in the table with mouse, the selection changes to the `TR`, resulting in the loss of our original selection. Therefore, what I am doing is retrieving the original selection when the selection moves to the TR using `this.historyResetLatestComputedSelection(true) `, so we don't lose it.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 12 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2420]: I added this `else if` statement because when I set the selection in the `if` block, the selection stops getting updated even if my cursor moves to the next `td`. So, what I'm doing is setting the focus node at `ev.target`, ensuring that when I move the cursor to other `tds`, the selection gets updated accordingly.

As for the conditions, firstly, I'm checking if the event is available, as I've observed that only `mousemove` passes the event when calling this function. This check is necessary because we only need to set the selection when selecting the table with the mouse. and I've included the second condition because in Firefox, when selecting with the mouse, it only selects within that table, and if the cursor goes out of that table, it stops updating the selection at that point.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 13 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2419]: But if the original selection was saved, that means at one point we had the correct selection. What changed it? Is it the browser or something in our code?

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 14 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2421]: > In Firefox, the selection shifts to the tr when selecting mutiple tds

The original code here took into account the fact that Firefox supports multiple ranges on a selection. So, if I remember correctly, when selecting multiple tds, Firefox has multiple ranges that span the separate cells and rows. The original code made a new single range by taking the full span of these multiple ranges. Isn't that ok anymore?

Also, to avoid infinite loops for sure, I believe you can avoid using `setSelection` by simply creating a range for your computations rather than actually setting the document's selection.

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2421]: This doesn't address our initial issue because even if we create a new single range in the `_handleSelectionInTable` function, our selection still remains on `tr`, and the enter process occurs based on this selection. So when Enter is pressed, it triggers `applyRawCommand`, within which `getDeepRange` sets the selection to the first p of the last row, and after that, `oEnter` is executed based on this position

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 13 above)
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2419]: Yes, until the second `td` is selected, we have the correct selection. However, when the second `td` is selected, the selection shifts to the `tr`. I believe this behavior is caused by the browser and is not related to our code.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 14 above)
Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2421]: I'm wondering if that's not the issue in the first place: that `oEnter` should do something special when it comes to tables so it takes the custom table selection (the classes) in consideration, and ignores the browser selection in the table. What do you think @dmo-odoo ?

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2421]: @Zynton Well I have a gut feeling that enter should behave differently in the particular case of tables, yes. The reason I think that is because I'm confused about what is supposed to happen when selecting multiple cells and press ENTER? I mean, it can't add a paragraph or anything, and it's nothing like BACKSPACE which is expected to empty the cells so... maybe it should do nothing at all? If selected multiple TDs then return ? I checked in GDocs but their behavior feels arbitrary and, frankly, buggy. cc @dhba-odoo

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2421]: That would be a decision for @pierre-pde, wouldn't it?

Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2421]: Just had a discussion with Pierre. The answer is: Only the cell that was the anchor for the start of the selection should be modified, ignoring the special table selection. The content of the anchor cell should be wiped out and end up with `<p><br></p><p><br>[]</p>`.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 15]
```diff
@@ -3457,11 +3457,19 @@ export class OdooEditor extends EventTarget {
...
                 this.historyRollback();
                 ev.preventDefault();
-                if (this._applyCommand('oEnter') === UNBREAKABLE_ROLLBACK_CODE) {
+                const anchorTD = closestElement(newSelection.anchorNode, '.o_selected_td');
+                const focusTD = closestElement(newSelection.focusNode, '.o_selected_td');
+                if (anchorTD && focusTD && closestElement(anchorTD, 'table') === closestElement(focusTD, 'table')) {
+                    this.deselectTable();
+                    setSelection(anchorTD.firstChild, 0, anchorTD.lastChild, nodeSize(anchorTD.lastChild));
+                }
+                if (ev.type === 'insertLineBreak') {
```
Comment (deso-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3468]: Same as above I think this should be `ev.inputType`.

File: addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js [hunk 16]
```diff
@@ -3903,6 +3903,143 @@ X[]
...
+                            <tbody>
+                                <tr>
+                                    <td><p><br></p><p>[]<br></p></td>
+                                    <td><p><br></p></td>
+                                </tr>
+                            </tbody>
+                        </table>`,
+                    ),
+                });
+            });
```
Comment (deso-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/test/spec/editor.test.js:4042]: Can we have a test case for `insertLineBreak` too.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js [hunk 17]
```diff
@@ -790,16 +790,15 @@ export function getDeepRange(editable, { range, sel, splitText, select, correctT
...
         (start !== end || startOffset !== endOffset) &&
-        (!beforeEnd || (beforeEnd.nodeType === Node.TEXT_NODE && !isVisibleStr(beforeEnd)))
+        (!beforeEnd || (beforeEnd.nodeType === Node.TEXT_NODE && !isVisibleStr(beforeEnd))) &&
+        previous && !closestElement(previous, 'table') && closestElement(previous).isContentEditable
     ) {
-        const previous = previousLeaf(endLeaf, editable, true);
-        if (previous && closestElement(previous).isContentEditable) {
-            [end, endOffset] = [previous, nodeSize(previous)];
-        }
+        [end, endOffset] = [previous, nodeSize(previous)];
```
Comment (deso-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:801]: Is this still required with the new specs?

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/utils/utils.js:801]: Yes, it's still required.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 14 above)
Comment (deso-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2431]: @Zynton As you mentioned that `oEnter` should do something special so it takes the custom table selection (the classes) rather than the browser's selection. However, with the new specs by @dmo-odoo, we considered setting the selection to the first `querySelector` value of `o_selected_td`. The issue with this approach is that if the selection is backward, the anchor node wouldn't be the first `o_selected_td`. We coundn't find another method to determine the correct starting point, so we are still reliant on the (second commit) browser's selection.

Comment (Zynton) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2431]: Yes I guess with this new spec once you've determined the selection is across cells but contained within one table, you'd have to use the browser selection to determine which cell to modify (but you have to consider the whole cell selected, which the browser might not).

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 18]
```diff
@@ -3375,11 +3373,19 @@ export class OdooEditor extends EventTarget {
...
+                const focusTD = closestElement(newSelection.focusNode, '.o_selected_td');
+                if (anchorTD && focusTD && closestElement(anchorTD, 'table') === closestElement(focusTD, 'table')) {
+                    this.deselectTable();
+                    setSelection(anchorTD.firstChild, 0, anchorTD.lastChild, nodeSize(anchorTD.lastChild));
+                }
+                if (ev.inputType === 'insertLineBreak') {
+                    this._applyCommand('oShiftEnter');
+                } else if (this._applyCommand('oEnter') === UNBREAKABLE_ROLLBACK_CODE) {
                     this._applyCommand('oShiftEnter');
                 }
```
Comment (deso-odoo) [inline addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3390]:
```suggestion
                if (
                    ev.inputType === 'insertLineBreak' ||
                    this._applyCommand('oEnter') === UNBREAKABLE_ROLLBACK_CODE
                ) {
                    this._applyCommand('oShiftEnter');
                }
```

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 13 above)
Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2365]: This requires a comment in the code because there is no way to guess this particular case just by reading the code.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js (hunk 14 above)
Comment (dmo-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2368]: Since the comment is still displayed, which means the line has not been changed, and the comment thread has not been marked resolved, I have a feeling the last comment of Zynton has not been applied, in particular the part about the whole cell being selected. That being said, when I try on runbot it seems to work. What am I missing? Because I don't understand how `setSelection(selection.anchorNode, selection.anchorOffset, ev.target, 0)` would in fact select the whole TD?

Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2367]: In Firefox, multiple ranges cause the original selection to be lost. To address this issue and preserve the original selection, I have created this commit. It is not related to single cell selection.
If you are referring to the single cell where the enter process will occur, then it will be at this [line](https://github.com/odoo/odoo/pull/139312/commits/405781d5ed9e484e9d6936054e591c2c575a2d05#diff-c309d19284e2e1f38a1dba1d68e0262371686595f3668a3bc0f1ca35e3189a0bR3413) in first commit

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 19]
```diff
@@ -2360,22 +2360,20 @@ export class OdooEditor extends EventTarget {
...
-            range = getDeepRange(this.editable, { correctTripleClick: anchorNode && anchorNode.nodeName === 'TR' });
+        if (this.isFirefox) {
+            if (selection.rangeCount > 1) {
+                this.historyResetLatestComputedSelection(true);
+            } else if (ev && closestElement(ev.target, 'table') === closestElement(selection.anchorNode, 'table') &&
+                    closestElement(selection.focusNode, 'td') !== closestElement(ev.target, 'td')) {
+                setSelection(selection.anchorNode, selection.anchorOffset, ev.target, 0);
+            }
         }
+        // We need the triple click correction only for a bug in firefox
```
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:2370]: Yes, we need to call getDeepRange for Chrome also. I don't see any issue with the comment, as it indicates that a triple-click is only required when the selection moves to a TR, which, as far as I know, only occurs in Firefox.

However, we can structure the code like this to avoid confusion and can be able to remove comment:

```
let range;
if (this.isFirefox) {
    if (selection.rangeCount > 1) {
        this.historyResetLatestComputedSelection(true);
    } else if (
        ev &&
        closestElement(ev.target, 'table') === closestElement(selection.anchorNode, 'table') &&
        closestElement(ev.target, 'td') !== closestElement(selection.focusNode, 'td')
    ) {
        setSelection(selection.anchorNode, selection.anchorOffset, ev.target, 0);
    }
    range = getDeepRange(this.editable, { correctTripleClick: anchorNode && anchorNode.nodeName === 'TR' });
} else {
    range = getDeepRange(this.editable);
}

```

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 20]
```diff
@@ -2360,22 +2360,20 @@ export class OdooEditor extends EventTarget {
...
-            // where it gives a selection of a full cell as tr 0 tr 1. The
-            // correction makes it so it gives us the cell and not its neighbor.
-            // In all other cases we don't want to make that correction so as to
-            // avoid flicker when hovering borders.
-            range = getDeepRange(this.editable, { correctTripleClick: anchorNode && anchorNode.nodeName === 'TR' });
+        if (this.isFirefox) {
+            if (selection.rangeCount > 1) {
+                this.historyResetLatestComputedSelection(true);
+            } else if (ev && closestElement(ev.target, 'table') === closestElement(selection.anchorNode, 'table') &&
+                    closestElement(selection.focusNode, 'td') !== closestElement(ev.target, 'td')) {
```
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: I have implemented the requested changes.

File: addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js [hunk 21]
```diff
@@ -3400,12 +3400,19 @@ export class OdooEditor extends EventTarget {
...
                 this.historyRollback();
                 ev.preventDefault();
-                if (this._applyCommand('oEnter') === UNBREAKABLE_ROLLBACK_CODE) {
+                getDeepRange(this.editable, { select: true, correctTripleClick: true });
+                const anchorTD = closestElement(newSelection.anchorNode, '.o_selected_td');
+                const focusTD = closestElement(newSelection.focusNode, '.o_selected_td');
+                if (anchorTD && focusTD && closestElement(anchorTD, 'table') === closestElement(focusTD, 'table')) {
+                    this.deselectTable();
+                    setSelection(anchorTD.firstChild, 0, anchorTD.lastChild, nodeSize(anchorTD.lastChild));
+                }
```
[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:3433]: @dmo-odoo The table breaks when _applyRawCommand is called for deleteRange, and this occurs before enter.js is executed. Therefore, the issue is not within enter.js but happens earlier, so moving the code to enter.js will not solve the issue. We could consider moving it to _applyRawCommand or deleteRange, but we need to ensure that the code only executes when the Enter key is pressed, and not for other keys.

[the remark this answers is not in the document]
Comment (dhba-odoo) [reply addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js]: @dmo-odoo done!

**Comment by dmo-odoo** [comment]
@dhba-odoo Thank you!
@robodoo rebase-ff r+
